crates/uv-requirements/src/pyproject.rs

```diff
@@ -1,133 +1,520 @@
+                given = format!("{given}@{rev}");
+            }
+            if let Some(subdirectory) = &subdirectory {
+                url.set_fragment(Some(&format!("subdirectory={subdirectory}")));
+                given = format!("{given}#subdirectory={subdirectory}");
+            }
+            let url = VerbatimUrl::from_url(url).with_given(given);
+            let repository = url.to_url().clone();
+            RequirementSource::Git {
+                url,
+                repository,
+                reference,
+                subdirectory: subdirectory.map(PathBuf::from),
+            }
+        }
+        Source::Url { url, subdirectory } => {
+            if matches!(requirement.version_or_url, Some(VersionOrUrl::Url(_))) {
+                return Err(UvSourcesLoweringError::ConflictingUrls);
+            }
+            let url = VerbatimUrl::from_url(Url::parse(&url)?).with_given(url);
+            RequirementSource::Url {
+                url,
+                subdirectory: subdirectory.map(PathBuf::from),
+            }
+        }
+        Source::Path { path, editable } => {
+            if matches!(requirement.version_or_url, Some(VersionOrUrl::Url(_))) {
+                return Err(UvSourcesLoweringError::ConflictingUrls);
+            }
+            path_source(path, project_dir, editable)?
+        }
+        Source::Registry { index } => match requirement.version_or_url {
+            None => return Err(UvSourcesLoweringError::UnconstrainedVersion),
+            Some(VersionOrUrl::VersionSpecifier(version)) => RequirementSource::Registry {
+                version,
+                index: Some(index),
+            },
+            Some(VersionOrUrl::Url(_)) => return Err(UvSourcesLoweringError::ConflictingUrls),
+        },
+        Source::Workspace {
+            workspace,
+            editable,
+        } => {
+            if matches!(requirement.version_or_url, Some(VersionOrUrl::Url(_))) {
+                return Err(UvSourcesLoweringError::ConflictingUrls);
+            }
+            if !workspace {
+                todo!()
+            }
+            let path = workspace_packages
+                .get(&requirement.name)
+                .ok_or(UvSourcesLoweringError::UndeclaredWorkspacePackage)?
+                .clone();
+            path_source(path, project_dir, editable)?
+        }
+        Source::CatchAll { .. } => {
+            // This is better than a serde error about not matching any enum variant
+            return Err(UvSourcesLoweringError::InvalidEntry);
+        }
+    };
+    Ok(Requirement {
+        name: requirement.name,
+        extras: requirement.extras,
+        marker: requirement.marker,
+        source,
+    })
+}
+
+/// Convert a path string to a path section.
+fn path_source(
+    path: String,
+    project_dir: &Path,
+    editable: Option<bool>,
+) -> Result<RequirementSource, UvSourcesLoweringError> {
+    let path_buf = PathBuf::from(&path);
+    let path_buf = path_buf
```

Comment:
I think you should be using `VerbatimUrl::parse_path` here? Why not? Why parse the URL out-of-band like this?

Reply:
Thanks, didn't realize that method exists